File: 프로그래머스/Level 2/더 맵게.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    
    priority_queue<int, vector<int>, greater<int>> q;
    
    for (int i = 0; i < scoville.size(); i++)
        q.push(scoville[i]);
    
    while (q.size() > 1) {
        if (q.top() >= K)
            break;
        
        int a = q.top();
        q.pop();
        int b = q.top();
        q.pop();
        
        int tmp = a + (b * 2);
        
        q.push(tmp);
        
        answer++;
    }
    
    if (q.size() == 1) {
        if (q.top() >= K)
            return answer;
        else
            return -1;
    }
    
    return answer;
}

File: 프로그래머스/Level 2/더 맵게.java
import java.util.*;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        if (scoville.length <= 1)
            return -1;
        
        PriorityQueue<Integer> q = new PriorityQueue<>();
        
        for (int i : scoville)
            q.add(i);
        
        while (!q.isEmpty()) {
            if (q.peek() >= K || q.size() == 1)
                break;
            
            int first = q.poll();
            int second = q.poll();
            
            q.add(first + (second * 2));
            answer++;
        }
        
        if (q.peek() < K)
            return -1;
        
        return answer;
    }
}
